### PR TITLE
Refactor vLLM server to use lm-format-enforcer when grammar sampling is enabled

### PR DESCRIPTION
**Pull request by jeffreymeetkai**
image

### CODE REVIEW

File: functionary/inference.py [hunk 1]
```diff
@@ -1,12 +1,19 @@
...
     LlamaForCausalLM,
     LlamaTokenizer,
     StoppingCriteria,
     StoppingCriteriaList,
 )
+from vllm.model_executor.guided_decoding.lm_format_enforcer_decoding import (
+    _cached_build_vllm_token_enforcer_tokenizer_data,
+    _normalize_json_schema_object,
+)
+from vllm.sampling_params import LogitsProcessor
```
Comment (khai-meetkai) [inline functionary/inference.py:16]: This import works only for new vllm version, right ?

File: functionary/inference.py [hunk 2]
```diff
@@ -1,12 +1,19 @@
 from typing import Dict, List, Optional, Union
 
 import torch
+from lmformatenforcer import CharacterLevelParser, JsonSchemaParser
+from lmformatenforcer.integrations.vllm import build_vllm_logits_processor
```
Comment (khai-meetkai) [inline functionary/inference.py:5]: lmformatforcer only works for server_vllm, right ? because server.py also uses this file so I am not sure if we should put the code for handling the logits processor in this file or not

File: functionary/inference_stream.py [hunk 3]
```diff
@@ -5,13 +5,19 @@
 import torch
 from transformers import LlamaForCausalLM, LlamaTokenizer
 from transformers.generation.logits_process import (
-    LogitsProcessorList, RepetitionPenaltyLogitsProcessor,
-    TemperatureLogitsWarper, TopKLogitsWarper, TopPLogitsWarper)
+    LogitsProcessorList,
```
Comment (khai-meetkai) [inline functionary/inference_stream.py:8]: The same here, server.py also uses this file so I am not sure we should put thing from vllm to this file

File: functionary/prompt_template/prompt_template_v2.py [hunk 4]
```diff
@@ -51,9 +43,12 @@ def get_predefined_function_names(self, function_types: Any) -> List[str]:
     def initialize_grammar_sampling_gen_state(
         self, tool_choice: str, curr_text: str, curr_tokens: List[int]
     ) -> Dict:
-        if tool_choice != "":
+        if tool_choice == "none":
             add_predefined_fns = False
-            stage = "pre-parameter"
+            stage = "no-tool-call"
+        elif tool_choice != "":
```
Comment (khai-meetkai) [inline functionary/prompt_template/prompt_template_v2.py:49]: Should add a comment here: force a function call

File: functionary/prompt_template/prompt_template_v2.py [hunk 5]
```diff
@@ -51,9 +43,12 @@ def get_predefined_function_names(self, function_types: Any) -> List[str]:
...
+            add_predefined_fns = False
+            stage = "parameter"
         else:
```
Comment (khai-meetkai) [inline functionary/prompt_template/prompt_template_v2.py:52]: Should add a comment indicate that this is for tool_choice=auto

**Review by khai-meetkai — APPROVED** [review]
